Decouple filter selection from channel construction

ctiller opened the pull request:
Allow plugins to extend the set of filters used by gRPC core:
- plugins at construction time can register against the 'channel_init' system to be allowed to mutate a new channel_stack_builder type
- channel_stack_builder provides a central and rather dynamic place to construct the list of filters required by a channel stack
- ultimately we construct the channel stack in the fashion we always have

This is also a prerequisite step to allowing filters to be implemented from wrapped languages.



@vjpai @sreecha @yang-g @nicolasnoble @jboeuf - I'd be interested in thoughts on the channel_init, channel_stack_builder interfaces. These will (likely) one day form a public but secondary/advanced interface to gRPC Core, and so API review is important.

dgquintas commented on src/core/channel/channel_stack_builder.c:
```diff
@@ -0,0 +1,262 @@
+
+int grpc_trace_channel_stack_builder = 0;
+
+#define BEGIN_SENTINAL ((grpc_channel_filter *)1)
```
s/sentinal/sentinel/g ?

dgquintas commented on src/core/channel/channel_stack_builder.h:
```diff
@@ -0,0 +1,140 @@
+void grpc_channel_stack_builder_iterator_destroy(
+    grpc_channel_stack_builder_iterator *iterator);
+
+// Destroy the builder, return the freshly minted channel stack
```
docstrings for all the non-trivial arguments?

dgquintas commented on src/core/surface/channel_init.h:
```diff
@@ -0,0 +1,76 @@
+ *
+ */
+
+#ifndef GRPC_INTERNAL_CORE_CHANNEL_CHANNEL_INIT_H
```
GRPC_INTERNAL_CORE_CHANNEL_CHANNEL_INIT_H isn't consistent with the file's full path

dgquintas commented on src/core/surface/channel_init.h:
```diff
@@ -0,0 +1,76 @@
+// It also provides a universal entry path to run those mutators to build
+// a channel stack for various subsystems.
+
+// One stage of mutation: call channel stack builder's to influence the finally
```
The genitive (builder's) seems to be dangling.

dgquintas commented on src/core/surface/channel_init.h:
```diff
@@ -0,0 +1,76 @@
+void *grpc_channel_init_create_stack(
+    grpc_exec_ctx *exec_ctx, grpc_channel_stack_type type, size_t prefix_bytes,
+    const grpc_channel_args *args, int initial_refs, grpc_iomgr_cb_func destroy,
+    void *destroy_arg, grpc_transport *optional_transport);
```
the last argument in the .c file is called "transport", without the "optional_" prefix. Is it optional? This argument is also  not part of the call to `grpc_channel_stack_builder_finish()` performed internally. It'd also be a good idea to document all other arguments regardless, IMHO.

dgquintas commented on src/core/surface/channel_init.h:
```diff
@@ -0,0 +1,76 @@
+// types.
+void grpc_channel_init_register_stage(grpc_channel_stack_type type,
+                                      int priority,
+                                      grpc_channel_init_stage stage,
```
s/stage/stage_fn? Somehow indicate the argument it's a "verb", not a "noun".

dgquintas commented:
Some docstring related nags. LGTM otherwise (+ the need to merge).

ctiller commented:
Alright, think I've addressed all the review stuff so far.
Merged with head.

I'd like to hold off merging this for a few days: @sreecha @vjpai @nicolasnoble @yang-g @dklempner @a-veitch @bogdandrutu @jboeuf @jtattermusch @nathanielmanistaatgoogle - can you think through any use cases you've seen for configuring things that look like grpc's filters, and check if those use cases are implementable within this framework (or trivial extensions of it). 

channel_init.h and channel_stack_builder.h are the primary interfaces of interest.

If there's nothing unresolvable, then let's merge on Wednesday.

dgquintas commented:
Remerge plz

ctiller commented:
Merged

dgquintas commented:
Sanity is failing due to nanopb: I think you need to do a submodule update to fetch the updated nanopb commit.

ctiller commented:
Merged with head again.

ctiller commented:
Fixed formatting; merge with head.

ctiller commented:
CC @makdharma

jtattermusch commented:
some tests still failing.

ctiller commented:
Regenerated build files, let's see what happens.
